[wip] Search pills

shubham-padia opened the pull request:
![peek 2018-06-08 23-38](https://user-images.githubusercontent.com/13910561/41173603-3209eaf0-6b75-11e8-986c-d9385dd519b7.gif)


Immediate TODO:
- Fix search icon overlap
- Add option of adjusting pill size in input_pill.js.
- Fix/(remove if required) the failing casper tests.
- Bug where an additional pill is appended when pressing enter, this is due to he text being not cleared before the code handling fr enter key occurs

timabbott commented:
@showell can you take this review?

showell commented:
There's a typo in one of the commit messages--search_pil.js.

showell commented on static/js/search.js:
```diff
@@ -63,7 +70,13 @@ exports.initialize = function () {
+        updater: function (search_string) {
+            // Order is important here. narrow_or_search_for_term
+            // assumes that the pill for the selected typeahead has
+            // not been added.
```
Can you explain why that's the case?  If I'm just reading this comment, I don't really know why `narrow_or_search_for_term` has that restriction, and maybe we can figure out a way so they don't depend on each other?  (I think the main thing is just to add to the comment.)

shubham-padia replied:
narrow_or_search_for_term gets a search string from existing pills and obtains existing operators. Newly selected suggestion is added to those operators. If narrow_or_search_for_term was called after append_search_string, the existing search pills at the time for calling that function would also have the newly selected suggestion, and appending it again would cause duplication. 
I've also added above text to the comment.

showell commented:
I did a first pass on this, and the overall code seems really clean.  I made a couple small comments, and the build seems to have failed.

showell commented:
@shubham-padia What is the status of your TODO list in the top comment here?  Is there anything we should hold off on merging?

shubham-padia commented:
@showell I was out for today, I'll get to the todo list tommorow. The first four commits should be good to merge. The commits after that start making changes to the UI, it would be best to merge them once the UI bugs are fixed.

shubham-padia commented:
@showell I've pushed the following changes:
1.) 3 of the 4 casper tests were failing as $.val() cannot be used for contenteditable divs, $.html() is used in `select_item_via_typeahead` instead. https://github.com/zulip/zulip/pull/9716/commits/dedb182294eb6ca8a59c6bad0b2752dc7edc6756 incorporates that change. This was was hard to figure out in particular. I was under the impression there was something wrong with the implementations. After trying multiple approaches, the thing that worked for was breaking the function being evaluated in `select_item_via_typeahead` and taking screenshots in between. I the noticed that no input was being inserted by $.val()
2.) Placing the logic for populating search pills in hashchange.js did not work as expected. `narrow.activate` calls by top left corner weren't removed once added and one from the stream list were not populated at all. The pre population logic was moved to narrow.activate. Pills are not populated
if the narrow was triggered by search as search handles the addition and removal of pill by itself. The reason for not handling search too in narrow.activate is to avoid clearing the pills and repopulating them.
Example of some of the triggers for narrow.activate include `restore draft`, `topic change`,`sidebar`.
3.) When pressing enter, two pills were added, one for the item you selected and one `search for x` where x is the input in the input box. That was fixed in 

The failure mentioned in point no.2 exists from https://github.com/zulip/zulip/pull/9716/commits/dedb182294eb6ca8a59c6bad0b2752dc7edc6756 to https://github.com/zulip/zulip/pull/9716/commits/b92bc4a3946610929fd8c05bef81444bc9f77179. It is fixed in https://github.com/zulip/zulip/pull/9716/commits/f84b78ae9595e68db66248c61e154040dbc87d48. It was not fixable in a single commit and I did not change the history in that case

shubham-padia commented:
@showell I'm curious about the merging policy here. I'm presuming that we will be merging this once the UI bugs are fixes and the UI seems stable. Is that correct?

showell commented:
@shubham-padia That's correct.  If you can set up your branch so that the first N commits are no-brainer merges (i.e. simple refactorings that pass all tests), I would consider opening a separate PR for them called something like "Add prep commits for search pills" and clearly link back to this PR (which already has "wip" in the title--thanks for doing that).

shubham-padia commented:
Cool :) Although the first 4 commits are mergeable, I'll let them be as it is a file addition and there I see no advantage in merging it beforehand.

shubham-padia commented:
Update: UI changes have been made and are under review on https://chat.zulip.org/#narrow/stream/101-design/topic/search.20pills .
A behavioural TODO is:
- taking account of the existing pills for the next suggestion. The current search takes care of that. (e.g if you have a stream filter applied, it will suggest that stream's topic)

shubham-padia commented:
Changes since the last update:
- https://github.com/zulip/zulip/pull/9716/commits/db3c7e0bea50e8122951565bbd400e2248d65644 Use search strings as the display value for search pills.
- https://github.com/zulip/zulip/pull/9716/commits/35c5091e167faf69ddc63b37d665c8ef1c3a1692 Fix search icon shifting left along with the first pill on overflow.
- https://github.com/zulip/zulip/pull/9716/commits/808accdbac80b555621b31396390102b4cda1865 Fix both cross icon overlap and its shifting left bug.

@showell @timabbott I've made the changes from the feedback on czo and some other irregularities I encountered. It'd be great to have a review on this. I feel this is a fairly stable experience for now and any other 
n czo and some other irregularities I encountered. It'd be great to have a review on this. I feel this is a fairly stable experience for now and any other UI/behavourial enhancements should be opened as a follow up.

Final design:
![peek 2018-06-19 04-02](https://user-images.githubusercontent.com/13910561/41565723-98ad7f9e-7375-11e8-9465-0907bf0623d9.gif)

shubham-padia commented:
Whenever the cross icon was not visible i.e empty search bar out
of focus, the left border of the search bar would appear as if it
was shortened. https://github.com/zulip/zulip/pull/9716/commits/8e4c4b2d06cf0950f234ef9d897c86f852c7f038 fixes that. 
Before:
![peek 2018-06-19 04-25](https://user-images.githubusercontent.com/13910561/41566677-297d19fe-737a-11e8-8a1c-752750fbb974.gif)

After:
![peek 2018-06-19 04-26](https://user-images.githubusercontent.com/13910561/41566687-3a459a7c-737a-11e8-8221-696a4fe7f81f.gif)

shubham-padia commented:
pinging @showell for review. (I think the previous request for review may have been buried in the multiple emails, please ignore otherwise :smiley: )

showell commented:
In FF I see pills like this (too low in search bar):

![image](https://user-images.githubusercontent.com/142908/42094795-e32a07b6-7b7e-11e8-8b47-9308fc69b357.png)

showell commented:
I guess completion now works one pill at a time?  It's actually kind of nice for some situations.

This seems to prevent negated search, or at least make it hard to edit.

showell commented:
Here's another strange situation you can get into:

![image](https://user-images.githubusercontent.com/142908/42095088-a0ec813e-7b7f-11e8-9af7-1be8b4aa8a64.png)

showell commented:
If I type "is:private", that turns into a pill, and so far, so good.  Then I type "V" and I'm presented with "stream Verona" as an option, which seems wrong, since "is:private" and "stream:Verona" are mutually exclusive options.

shubham-padia commented:
> If I type "is:private", that turns into a pill, and so far, so good. Then I type "V" and I'm presented with "stream Verona" as an option, which seems wrong, since "is:private" and "stream:Verona" are mutually exclusive options.

Yes, the suggestions do not take previous pills into account right now. I was hoping to do that once the current code gets reviewed. 
I had mentioned in the behavioural todo above and was going to mention it in the review request, but it got out of my mind :sweat_smile:.

timabbott commented:
Thanks for working on this @shubham-padia!  After we reworked this to have all the changes beneath a settings.SEARCH_PILLS_ENABLED setting, this is now nearly ready to merge (I pushed an updated version here).  There's a couple small fixes we should do before merging:
* "quiet" feature for input pills should be its own commit.
* Casper tests are broken

Once we're done, we can work through this list of TODO items that need to be resolved before it can be merged:
* Rename my_pill to a clearer name for the group of all search pills
* / hotkey doesn't work
* Escape hotkey to unnarrow doesn't work if search is focused
* After backspace, don't produce an (empty) typeahead
* backspace should remove a search pill even when you have typeahead open
* (styling) Remove line to the right of magnifying glass
* (styling) Visually show that we're focused on the search area (e.g. a colored border; the existing 
![image](https://user-images.githubusercontent.com/2746074/42723093-1f30306c-8775-11e8-9cec-32b8487b96be.png) should be fine
) 
* (styling) The X at the right end of the search area is misplaced.

* (medium priority) Don't allow creating a pill when validation doesn't pass
* (low priority) tab should open typeahead for where you are?

shubham-padia commented:
@timabbott Thanks for helping with the rework. I've fixed the casper tests and made `quiet` feature for input pills its own commit.

timabbott commented:
I made one change to ef2f6b7e47cf157e78dcf58176534eed80081fcf, which is to add a `return true` for the contenteditable case, and merged this.  Thanks @shubham-padia!  

I also updated the checklist of follow-up items, above, based on your comment.  Can you move the remaining to a new issue, that we use as the "remaining search pills implementation details" TODO list?  This PR has a long enough history that it's not a great place to keep that.  

One other detail worth mentioning: Changes @synicalsyntax made to the user pills ended up breaking some of the styling for this; so you two will want to collaborate to get the search pills styling correct again.

I'm looking forward to this!  Aside from the few annoying quirks, this feels a lot nicer :)

shubham-padia commented:
Open #10026 for follow-up issues
